src/config.c

```diff
@@ -840,40 +858,30 @@ void configSetCommand(client *c) {
         }
 
         /* If this config appears twice then fail */
-        sds config_name = sdsnew(config->name);
-        if (dictFind(set_configs, config_name)) {
+        if (dictFind(set_config_values, config->name)) {
             /* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
             errstr = "duplicate parameter";
             err_arg_name = c->argv[2+i*2]->ptr;
             invalid_args = 1;
         } else {
-            dictAdd(set_configs, sdsnew(config->name), config);
-            dictAdd(new_values, sdsnew(config->name), c->argv[2+i*2+1]->ptr);
+            config_values[i].config = config;
+            config_values[i].old_value = config->interface.get(config);
+            config_values[i].new_value = c->argv[2+i*2+1]->ptr;
+            dictAdd(set_config_values, sdsnew(config->name), &config_values[i]);
         }
-        sdsfree(config_name);
         config_names[i] = config->name;
     }
     
     if (invalid_args) goto err;
 
-    /* Backup old values before setting new ones */
-    dictIterator *di = dictGetIterator(set_configs);
-    dictEntry *de;
-    while ((de = dictNext(di)) != NULL) {
-        standardConfig *config = dictGetVal(de);
-        dictAdd(old_values, sdsnew(config->name), config->interface.get(config));
-    }
-    dictReleaseIterator(di);
-
     /* Set all new values (don't apply yet) */
     for (i = 0; i < config_count; i++) {
-        sds config_name = sdsnew(config_names[i]);
-        standardConfig *config = dictGetVal(dictFind(set_configs, config_name));
-        sds new_value = dictGetVal(dictFind(new_values, config_name));
-        sdsfree(config_name);
+        configSetValue *config_value = dictGetVal(dictFind(set_config_values, config_names[i]));
+        standardConfig *config = config_value->config;
+        sds new_value = config_value->new_value;
```

Comment:
```suggestion
        configSetValue *config_value = dictGetVal(dictFind(set_config_values, config_names[i]));
        standardConfig *config = config_value->config;
```
I think we can just use `config_value->new_value` directly.